feat: add abort action storage

jakubmkowalski opened the pull request:
Closes #345
This pull request introduces significant changes to support an "Abort Action" feature for transactions in progress. It includes new functionality for aborting transactions, validation logic, and updates to transaction management methods. Additionally, several refactorings and enhancements were made to improve code structure and testing capabilities.

### New Abort Action Feature:
* [`pkg/internal/fixtures/default_abort_action.go`](diffhunk://#diff-a531be6724031e5b6929102591c4cddbdc42343a00d5210d947e970999ffac58R1-R45): Added default test fixtures for `AbortActionArgs` and `AbortActionResult` to facilitate testing of the abort feature.
* [`pkg/storage/internal/actions/abort.go`](diffhunk://#diff-9cda57e1464cf412f67e91af2cdd3eb6ff2de6f01548d230293df6d9788e0443R1-R91): Implemented the core logic for aborting transactions, including validation of transaction status and updates to transaction states.
* [`pkg/internal/validate/validate_abort_action.go`](diffhunk://#diff-23182a6f2da432dafd807f195c5c0b532a3ba400712bed5e733c2e3c6e0e421eR1-R19): Added validation logic for `AbortActionArgs` to ensure proper input for aborting transactions.

### Updates to Transaction Management:
* [`pkg/internal/storage/repo/transactions.go`](diffhunk://#diff-f4b526ed9ad7b3f1f6240c510c35ae64860702054261613ab2b51ef58eb8a4dcL319-R480): Refactored transaction status update methods to separate logic for updating by `txID` and internal `transactionID`. Added methods for cleaning up transaction outputs and releasing reserved UTXOs.
* [`pkg/storage/internal/actions/repos.interface.go`](diffhunk://#diff-94f32606e818054d510d6fadc9d60851e34ade46a455a66f8dee6f35f70f6a03R30-R52): Updated the `TransactionsRepo` interface to include new methods for finding transactions by reference and updating transaction status by ID.

### Mock and Client Enhancements:
* [`pkg/internal/mocks/mock_wallet_storage_writer.go`](diffhunk://#diff-d4a15600e36c87de294b6ce1f4a7e7407037840eed3f16d2fc5e596ae1532765R44-R58): Added mock methods for `AbortAction` to support testing of the abort feature.
* [`pkg/storage/client_gen.go`](diffhunk://#diff-43960e1c778173ecad4d9320004cfd58c26530c3e2dc8f012195aed885029b4aR93-R97): Added `AbortAction` method to the `WalletStorageProviderClient` for client-side interaction with the abort feature. [[1]](diffhunk://#diff-43960e1c778173ecad4d9320004cfd58c26530c3e2dc8f012195aed885029b4aR93-R97) [[2]](diffhunk://#diff-43960e1c778173ecad4d9320004cfd58c26530c3e2dc8f012195aed885029b4aR114)

### Testability Improvements:
* [`pkg/storage/internal/testabilities/assertions_db_state.go`](diffhunk://#diff-fc7a5f94d98dfb749d5b4fe804c082f5710837763c7a207efdc6aae5a90cacd5R55): Added a new assertion method `WithTxIDCount` for validating outputs associated with specific transactions in tests. [[1]](diffhunk://#diff-fc7a5f94d98dfb749d5b4fe804c082f5710837763c7a207efdc6aae5a90cacd5R55) [[2]](diffhunk://#diff-fc7a5f94d98dfb749d5b4fe804c082f5710837763c7a207efdc6aae5a90cacd5R246-R258)

### Minor Changes:
* [`pkg/internal/testabilities/dbfixtures/db.go`](diffhunk://#diff-8837b63d4f3f1ba1b8362ad079cbae01c917ebc6b80f249ec5d4e1798c9a4bcfL34-R35): Modified SQLite connection string for tests to remove memory mode.
## Description of Changes

Provide a brief description of the changes you've made.

## Linked Issues / Tickets

Reference any related issues or tickets, e.g. "".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

jakubmkowalski committed:
feat: implement abort action storage for transactions with validation and repository integration

jakubmkowalski committed:
feat: add abort action tests with defaults and mocks

jakubmkowalski committed:
feat: implement transaction status update and validation for abort actions

jakubmkowalski committed:
feat: enhance abort action by restoring inputs on transaction failure and updating tests

jakubmkowalski committed:
feat: refactor transaction status update methods and enhance abort action validation

jakubmkowalski committed:
feat: add abort action handling and update transaction status methods

jakubmkowalski committed:
refactor: rename test functions for consistency and clarity

jakubmkowalski committed:
fix: update error messages for abort action validation and tests

jakubmkowalski committed:
fix: update error messages for abort action and assertion tests

mgosek-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -225,6 +224,32 @@ func (txs *Transactions) FindTransactionByReference(ctx context.Context, userID
 	return txs.mapModelToTransactionEntity(&transaction), nil
 }
+func (txs *Transactions) FindUniqueTransactionByReference(ctx context.Context, userID int, reference string) (*entity.Transaction, error) {
+	transaction := []models.Transaction{}
```
You can also use zero value `var transactions []models.Transaction`

jakubmkowalski replied:
changed

mgosek-4chain commented on pkg/storage/internal/testabilities/assertions_db_state.go:
```diff
@@ -311,13 +312,26 @@ func (d *outputsListAssertion) WithCount(expected int) OutputsListAssertion {
+
+		equal := txIDStr == id
+		return equal
+	}))
```
Why not just use a simple for loop? 🤔

jakubmkowalski replied:
The implementation is similar to WithCountHavingOutpoint. 
If this is only a suggestion, I would leave it as it is, unless someone else is not satisfied with this implementation.

chris-4chain replied:
It seems ok to me - especially if similar thing is already in the code.

dorzepowski replied:
Look at my comment for whole method 😀

https://github.com/bsv-blockchain/go-wallet-toolbox/pull/395/files/e27c98813b962666439321096bd6d906c0da5415#r2221503488

jakubmkowalski replied:
method removed

jakubmkowalski committed:
fix: change initialization and rename variable for clarity in FindUniqueTransactionByReference method

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -315,32 +340,132 @@ func makeOutputsSpendable(tx *gorm.DB, updatedTx entity.UpdatedTx) error {
+	ctx context.Context,
+	transactionID uint,
+	newStatus wdk.TxStatus,
+	provenTxReqStatus wdk.ProvenTxReqStatus,
```
This `provenTxReqStatus` is not used. Please make sure if it's required (logically).
It not - remove from the args list

jakubmkowalski replied:
removed

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -315,32 +340,132 @@ func makeOutputsSpendable(tx *gorm.DB, updatedTx entity.UpdatedTx) error {
+		}).Error
+}
+
+func (txs *Transactions) updateTransactionStatusByInternalID(tx *gorm.DB, transactionID uint, newStatus wdk.TxStatus) error {
```
Why not `updateTransactionStatusByID` ?
- Why the `"internal"` word?

jakubmkowalski replied:
It is ID from database, i didn't want to use TransactionID nor TxID but you are right simple ID will be ok.
Changed to `updateTransactionStatusByID`

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -315,32 +340,132 @@ func makeOutputsSpendable(tx *gorm.DB, updatedTx entity.UpdatedTx) error {
+		Where("spent_by = ?", transactionID).
 		Updates(map[string]any{
-			"status": txStatus,
+			"spent_by":  nil,
```
Hmmmm - what if the output is already spent (`spent_by` is not nil)?

jakubmkowalski replied:
You are right, when `spent_by` in not `null` aborting should fail.

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+		return nil, err
+	}
+
+	history := history.NewBuilder().AbortAction(*args.Reference)
```
Maybe `historyNote` or `note` as a variable name, because `"history"` is already a package name

jakubmkowalski replied:
changed to `historyNote`

chris-4chain commented on pkg/internal/storage/history/history_note.go:
```diff
@@ -16,6 +16,8 @@ const (
 	ProcessActionHistoryNote     = "processAction"
 	AggregateResultsHistoryNote  = "aggregateResults"
+	AbortActionHistoryNote       = "abortAction"
+	ActionHistoryNote            = "action"
```
This `ActionHistoryNote` seems unused

jakubmkowalski replied:
removed

chris-4chain commented on pkg/internal/storage/history/history_note.go:
```diff
@@ -112,6 +118,12 @@ func (b *builder) PostBeefSuccess(serviceName string, beef []byte, txIDs []strin
 
+func (b *builder) AbortAction(reference string) Builder {
+	return b.WithWhat(AbortActionHistoryNote).
+		WithAttribute(actionAttr, abortActionAttr).
```
This is weird / inconsistent. 
Why additional `actionAttr`? We have already the information about "what action is made" in `"AbortActionHistoryNote"`

jakubmkowalski replied:
simplified to:
```
		return b.WithWhat(AbortActionHistoryNote).
		WithAttribute(referenceAttr, reference)
```

chris-4chain commented on pkg/storage/internal/actions/repos.interface.go:
```diff
@@ -28,6 +28,7 @@ type TransactionsRepo interface {
 	CreateTransaction(ctx context.Context, transaction *entity.NewTx) error
 	FindTransactionByUserIDAndTxID(ctx context.Context, userID int, txID string) (*entity.Transaction, error)
 	FindTransactionByReference(ctx context.Context, userID int, reference string) (*entity.Transaction, error)
+	FindUniqueTransactionByReference(ctx context.Context, userID int, reference string) (*entity.Transaction, error)
```
This seems redundant to `FindTransactionByReference`.
The second method is not necessary. 
However, if you think that there should be also checking of "uniquness" (like, `len(transactions) != 1`) - feel free to add it to `FindTransactionByReference`

jakubmkowalski replied:
I added this condition to the `FindTransactionByReference` function and removed `FindUniqueTransactionByReference`. 
I want to be sure that there is no more than one transaction with this reference. 
If you know that such a situation is impossible or very unlikely, I can remove this condition.

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+		return nil, fmt.Errorf("failed to find unique transaction by reference: %w", err)
+	}
+
+	if txEntity == nil && len(*args.Reference) == 64 {
```
It seems like magic - IMHO we should provide some `NOTE` comment or separate this logic to another function (also with the comment)

dorzepowski replied:
I would solve this with extracting a method to use here like
```suggestion
if txEntity == nil && a.isPotentiallyTxID(args.Reference) {
```

jakubmkowalski replied:
method extracted

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+	}, nil
+}
+
+func validateTxForAbort(txEntity *entity.Transaction) error {
```
it could be a private method for `(a *abortAction)`.
Then, the name could be shorter, like: `validateTx`

jakubmkowalski replied:
changed

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+
+func validateTxForAbort(txEntity *entity.Transaction) error {
+	if !txEntity.IsOutgoing {
+		return fmt.Errorf("txDirectionInvalid: must be an outgoing transaction.")
```
Warning from my IDE:
`Error string should not be capitalized or end with punctuation mark`.

Could you remove the period character `"."`?

jakubmkowalski replied:
removed

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+	return validateTxStatusForAbort(txEntity.Status)
+}
+
+func validateTxStatusForAbort(txStatus wdk.TxStatus) error {
```
I would suggest using a `switch` statement for this.
It is more performant:

```golang
func validateTxStatusForAbort(txStatus wdk.TxStatus) error {
	switch txStatus {
	case wdk.TxStatusCompleted, wdk.TxStatusFailed, wdk.TxStatusSending, wdk.TxStatusUnproven:
		return fmt.Errorf("txNotAbortable: action with status %s cannot be aborted", txStatus)
	case wdk.TxStatusUnprocessed, wdk.TxStatusUnsigned, wdk.TxStatusNoSend, wdk.TxStatusNonFinal, wdk.TxStatusUnfail:
		fallthrough
	default:
		return nil
	}
}
```

The second case is needed, because otherwise the linter screams "exhaustive".

... or you can disable linter-exhaustive for this switch:
```golang
	//nolint:exhaustive /
```

jakubmkowalski replied:
changed to switch case

ac4ch replied:
hmm .. what about doing it like this ? 

```go
var abortableStatuses = map[wdk.TxStatus]struct{}{
	wdk.TxStatusUnprocessed: {},
	wdk.TxStatusUnsigned:    {},
	wdk.TxStatusNoSend:      {},
	wdk.TxStatusNonFinal:    {},
	wdk.TxStatusUnfail:      {},
}

func validateTxStatusForAbort(txStatus wdk.TxStatus) error {
	if _, ok := abortableStatuses[txStatus]; ok {
		return nil
	}

	switch txStatus {
	case wdk.TxStatusCompleted, wdk.TxStatusFailed, wdk.TxStatusSending, wdk.TxStatusUnproven:
		return fmt.Errorf("%w: action with status %s cannot be aborted", wdk.ErrNotAbortableAction, txStatus)
	}

	return fmt.Errorf("%w: unexpected transaction status %s", wdk.ErrNotAbortableAction, txStatus)
}
```
i think is easier to see what can be aborted ... and what cannot be ... but that could be just me :)

chris-4chain replied:
I totally don't get above suggestion.

ac4ch replied:
sure but i do...

chris-4chain replied:
I don't understand why you mixed the `map` and `switch` approach.
I think, the switch is the most efficient and straightforward way to do so.
And thanks to lint-exhaustive we are sure that we cover all possible statuses.

jakubmkowalski replied:
Looking at your implementation, I would allocate two maps, one for statuses that can be aborted and another for statuses that cannot be aborted.
I prefer the implementation with maps, but I will follow Chris suggestion.

ac4ch replied:
because in my opinion is 
clear whitelist of what can be aborted.... 
proper error for unexpected statuses....
separation of known-non-abortables vs. unknown statuses....

as your suggestion you accepting unknown 

anyway up to Jakub what he prefer ... its just suggestion and i do not insist on implementing it

chris-4chain replied:
My original point and the main suggestion is:
- "do not employ maps when you can use switch-case"
- maps have its performance overhead 
- and in this situation I don't see any reason to do so.

chris-4chain replied:
Only thing, that could be adjusted (comparing to my initial suggestion) is to return error on "unknown" status, like:

```golang
func validateTxStatusForAbort(txStatus wdk.TxStatus) error {
	switch txStatus {
	case wdk.TxStatusCompleted, wdk.TxStatusFailed, wdk.TxStatusSending, wdk.TxStatusUnproven:
		return fmt.Errorf("txNotAbortable: action with status %s cannot be aborted", txStatus)
	case wdk.TxStatusUnprocessed, wdk.TxStatusUnsigned, wdk.TxStatusNoSend, wdk.TxStatusNonFinal, wdk.TxStatusUnfail:
		return nil
	default:
		return fmt.Errorf("unknown status %s", txStatus)
	}
}
```

jakubmkowalski replied:
Yes, I added error for "unknown" status earlier. So the implementation now looks like in your last message

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -315,32 +340,132 @@ func makeOutputsSpendable(tx *gorm.DB, updatedTx entity.UpdatedTx) error {
+	provenTxReqStatus wdk.ProvenTxReqStatus,
+) error {
+	err := txs.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
+		if newStatus == wdk.TxStatusFailed {
```
This function `UpdateTransactionStatusByID` mixes responsibilities. Actually, it doesn't "only" "update transaction status by id".

jakubmkowalski replied:
refactored

dorzepowski commented on pkg/wdk/storage_abort_action_args.go:
```diff
@@ -0,0 +1,13 @@
+// AbortActionArgs defines the arguments for aborting a wallet action.
+type AbortActionArgs struct {
+	// Reference is the unique identifier for the action to be aborted.
+	Reference *string
```
1. this doesn't seem to be optional in typescript, so it shouldn't be a pointer
2. In typescript validation it seems that it is required to be Base64 encoded, so you can use primitives.Base64String type to simplify validation later

jakubmkowalski replied:
1. Now `Reference` is not a pointer.
2. I would do this as separate task as this requires some more refactoring

dorzepowski commented on pkg/storage/provider.go:
```diff
@@ -505,3 +505,20 @@ func (p *Provider) CommissionEntity() crud.Commission {
 }
+
+// AbortAction aborts an action by its reference for the authenticated user.
+func (p *Provider) AbortAction(ctx context.Context, auth wdk.AuthID, args wdk.AbortActionArgs) (*wdk.AbortActionResult, error) {
```
please move it closer to the rest of the XyzAction methods (for example just after ProcessAction)

jakubmkowalski replied:
moved

dorzepowski commented on pkg/internal/validate/validate_abort_action.go:
```diff
@@ -0,0 +1,19 @@
+		return fmt.Errorf("args cannot be nil")
+	}
+
+	if args.Reference == nil {
```
Check for empty string.

Hint:
```suggestion
	if is.BlankString(args.Reference) {
```

jakubmkowalski replied:
as `Reference` is now type `primitives.Base64String` changed to:
```
if args.Reference == "" {
```

dorzepowski commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+
+func newAbortAction(logger *slog.Logger, transactions TransactionsRepo) *abortAction {
+	return &abortAction{
+		logger:           logging.Child(logger, "abort_action"),
```
### minor

We don't have strict convention (worth to talk about it later and adjust all), but in other XyzAction we use camel case for service name

```suggestion
		logger:           logging.Child(logger, "abortAction"),
```

jakubmkowalski replied:
changed

dorzepowski commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	if unAbortableStatuses[txStatus] {
+		return fmt.Errorf("txNotAbortable: action with status %s cannot be aborted", txStatus)
```
It looks like you had some intuition or something that you're trying to distinguish not abortable errors from others.

I would suggest to create a public error like:
```go
var ErrNotAbortableAction = fmt.Errorf("action cannot be aborted")
```


And then use it here:
```suggestion
		return fmt.Errorf("%w: action with status %s cannot be aborted", wdk.ErrNotAbortableAction,  txStatus)
```

jakubmkowalski replied:
changed

dorzepowski commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -315,32 +340,132 @@ func makeOutputsSpendable(tx *gorm.DB, updatedTx entity.UpdatedTx) error {
+
+func (txs *Transactions) releaseOutputsReservedByTransaction(tx *gorm.DB, transactionID uint) error {
+	return tx.Model(&models.Output{}).
+		Where("spent_by = ?", transactionID).
```
@chris-4chain can we somehow use gorm gen result for those new queries?

jakubmkowalski replied:
@chris-4chain bump

dorzepowski replied:
We talked about it, that it is possible, but we should introduce it in separate PR, because it's more change then I thought at the beginning.

dorzepowski commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,88 @@
+		}
+	}
+
+	err = a.transactionsRepo.UpdateTransactionStatusByID(ctx, txEntity.ID, wdk.TxStatusFailed, wdk.ProvenTxStatusInvalid)
```
maybe I miss something, but if the txEntity has TxID, then we're updating the status twice?

IMHO you should use separated method for UpdateKnownTxStatus in case TxID exists.

jakubmkowalski replied:
refactored

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+
+func TestAbortActionSuccess(t *testing.T) {
+	// given:
+	given, cleanup := testabilities.GivenCustomStorage(t, fixtures.StorageServerPrivKey, "dbstorage_test")
```
You don't need to use CustomStorage, it's used for synchronisation where you need 2 storages (look at the tests for CreateActions for example)

jakubmkowalski replied:
now using default db

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+		fixtures.DefaultValidCreateActionArgs(),
+	)
+	require.NoError(t, err)
+	testabilities.ThenDBState(t, activeStorage).AllOutputs(testusers.Alice).WithCount(1 + NumberOfDesiredUTXOs)
```
No need to check the createAction result here - the tests for create action should test it

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/internal/testabilities/assertions_db_state.go:
```diff
@@ -311,13 +312,26 @@ func (d *outputsListAssertion) WithCount(expected int) OutputsListAssertion {
 	return d
 }
 
+func (d *outputsListAssertion) WithTxIDCount(expected int, txID uint) OutputsListAssertion {
```
It looks like it is not used anywhere

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+	given.Faucet(activeStorage, testusers.Alice).TopUp(100_000)
+	createdTx, _ := given.ActionCreatedAndSigned(activeStorage)
+	txID := createdTx.Inputs[0].SourceTxID
+	testabilities.ThenDBState(t, activeStorage).AllOutputs(testusers.Alice).WithCount(1 + NumberOfDesiredUTXOs)
```
again, not needed check

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+
+	activeStorage := given.Provider().GORM()
+
+	given.Faucet(activeStorage, testusers.Alice).TopUp(100_000)
```
the top up here is not needed

jakubmkowalski replied:
now I use 
given.Faucet(activeStorage, testusers.Alice).TopUp(100_000) 
or 
given.ActionCreatedAndSigned(activeStorage)
but not both in single test

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+	activeStorage := given.Provider().GORM()
+
+	given.Faucet(activeStorage, testusers.Alice).TopUp(100_000)
+	createdTx, _ := given.ActionCreatedAndSigned(activeStorage)
```
Interesting - I would assume that result of this would be exactly the outgoing action, 
Maybe it's worth checking if this is not a bug.

@chris-4chain Was this intentional or we should create an issue to investigate

dorzepowski replied:
@chris-4chain nvm

Ohh, I see that I overlooked next line. So you just taking the parent tx.

I'm suggesting to either 
 use the txID from faucet topup
OR add some better spacing to better expose the line from where the txID is comming from.


Additionally, it seems that you can use this `given.ActionCreatedAndSigned(activeStorage)` in all the other tests, instead of plain call to CreateAction

jakubmkowalski replied:
Now I hope that the readability has improved.

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+	faucet := given.Faucet(activeStorage, testusers.Alice)
+	faucet.TopUp(100_000)
+
+	createResult, err := activeStorage.CreateAction(
```
It would be great to use 
`given.ActionCreatedAndSigned(activeStorage)` 
In most of the tests here.

jakubmkowalski replied:
now when I can, I use `given.ActionCreatedAndSigned(activeStorage)`

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+
+	// then:
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "failed to abort action: txDirectionInvalid: must be an outgoing transaction.")
```
If you will create this public error (as I'm suggesting in other comment), then here you could use require.ErrorIs

jakubmkowalski replied:
tests now looks little different, i don't know if this applies now

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+
+func TestAbortActionTransactionNotOutgoing(t *testing.T) {
+	// given:
+	given, cleanup := testabilities.GivenCustomStorage(t, fixtures.StorageServerPrivKey, "dbstorage_test")
```
again, no need for custom storage

jakubmkowalski replied:
now using default db

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,272 @@
+	require.Contains(t, err.Error(), "found 0")
+}
+
+func TestAbortActionTransactionStatusFailed(t *testing.T) {
```
The tests for different statuses could (and should) be changed to table tests

jakubmkowalski replied:
changed to table tests

dorzepowski commented on pkg/internal/fixtures/default_abort_action.go:
```diff
@@ -0,0 +1,45 @@
+	return &wdk.AbortActionResult{
+		Aborted: true,
+	}
+}
```
all those methods seem to be not used

jakubmkowalski replied:
removed unused

dorzepowski commented on pkg/internal/validate/validate_abort_action.go:
```diff
@@ -0,0 +1,19 @@
+	"github.com/bsv-blockchain/go-wallet-toolbox/pkg/wdk"
+)
+
+func ValidAbortActionArgs(args *wdk.AbortActionArgs) error {
```
We should have tests for that

jakubmkowalski replied:
tests added

jakubmkowalski committed:
Refactor abort action handling and validation

- Simplified AbortActionArgs structure by removing pointer for Reference.
- Introduced ValidAbortActionArgs function to validate abort action arguments.
- Added FindTransactionByTxID method to Transactions repo for fetching transactions by TxID.
- Implemented AbortTransactionAtomic method to handle transaction abort logic atomically.
- Updated abort action logic in the actions package to utilize new transaction methods.
- Enhanced test coverage for abort action scenarios, including various error cases.
- Removed unused code and comments to improve readability and maintainability.

jakubmkowalski committed:
Merge branch 'feat/add_abort_action_storage' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_abort_action_storage

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

jakubmkowalski committed:
feat: enhance abort transaction handling with history tracking

jakubmkowalski committed:
refactor: remove commented-out method from TransactionsRepo interface

jakubmkowalski committed:
refactor: remove unused import from repos.interface.go

jakubmkowalski committed:
refactor: replace FindUniqueTransactionByReference with FindTransactionByReference

jakubmkowalski committed:
refactor: remove unnecessary comment in HasUserTransactionByTxID method

ac4ch commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -485,3 +554,47 @@ func (txs *Transactions) labelFilterScope(tx *gorm.DB, userID int, filter entity
+	}
+
+	return nil
+}
```
```suggestion
func (txs *Transactions) AbortTransactionAtomic(ctx context.Context, transactionID uint, txID *string, reference string) error {
	if err := txs.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		txNotes := history.NewBuilder().AbortAction(reference).Note().AsList()
		historyBuilders := make([]history.Builder, len(txNotes))
		for i, note := range txNotes {
			historyBuilders[i] = history.NewBuilderFromNote(note)
		}

		steps := []func(*gorm.DB) error{
			func(tx *gorm.DB) error {
				return txs.releaseReservedUTXOs(tx, transactionID)
			},
			func(tx *gorm.DB) error {
				return txs.checkIfAnyOutputIsSpent(tx, transactionID)
			},
			func(tx *gorm.DB) error {
				return txs.releaseOutputsReservedByTransaction(tx, transactionID)
			},
			func(tx *gorm.DB) error {
				return txs.deleteOutputsByTransactionID(tx, transactionID)
			},
			func(tx *gorm.DB) error {
				return txs.updateTransactionStatusByID(tx, transactionID, wdk.TxStatusFailed)
			},
		}

		for _, step := range steps {
			if err := step(tx); err != nil {
				return err
			}
		}

		if txID != nil && *txID != "" {
			if err := updateKnownTxStatus(tx, *txID, wdk.ProvenTxStatusInvalid, historyBuilders); err != nil {
				return err
			}
		}

		return nil
	}); err != nil {
		return fmt.Errorf("failed to abort transaction: %w", err)
	}

	return nil
}
```

consider to improve the readability ... i think is slightly cleaner coz it reduces nesting by using the `if err`

jakubmkowalski replied:
I followed your suggestion with minor changes

ac4ch commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,83 @@
+
+	err = a.validateTx(txEntity)
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("transaction validation failed: %w", err)
```

jakubmkowalski replied:
changed

ac4ch commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,83 @@
+	return &wdk.AbortActionResult{
+		Aborted: true,
+	}, nil
+}
```
```suggestion
func (a *abortAction) AbortAction(ctx context.Context, userID int, args *wdk.AbortActionArgs) (*wdk.AbortActionResult, error) {
	txEntity, err := a.transactionsRepo.FindTransactionByReference(ctx, userID, args.Reference)
	if err != nil {
		return nil, fmt.Errorf("failed to find unique transaction by reference: %w", err)
	}

	if txEntity == nil && a.isPotentiallyTxID(args.Reference) {
		txEntity, err = a.transactionsRepo.FindTransactionByUserIDAndTxID(ctx, userID, args.Reference)
		if err != nil {
			return nil, fmt.Errorf("failed to find transaction by txid: %w", err)
		}
	}

	if txEntity == nil {
		return nil, fmt.Errorf("no transaction found with reference or txid %q", args.Reference)
	}

	if err := a.validateTx(txEntity); err != nil {
		return nil, fmt.Errorf("transaction validation failed: %w", err)
	}

	if err := a.transactionsRepo.AbortTransactionAtomic(ctx, txEntity.ID, txEntity.TxID, args.Reference); err != nil {
		return nil, fmt.Errorf("failed to abort transaction: %w", err)
	}

	return &wdk.AbortActionResult{Aborted: true}, nil
}
``` 
or maybe like this WDYT ?

jakubmkowalski replied:
I applied your suggestion

jakubmkowalski committed:
refactor: streamline error handling and improve transaction abort logic

jakubmkowalski committed:
refactor: enhance error handling in AbortTransactionAtomic method to remove linter issues

jakubmkowalski committed:
fix: adjust expepected error msgs in tests

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

dorzepowski commented on pkg/internal/storage/history/history_note.go:
```diff
@@ -26,12 +27,16 @@ const (
 	statusNowAttr   = "status_now"
 	serviceNameAttr = "name"
+	abortActionAttr = "abort_action"
+	actionAttr      = "action"
```
you forgot to remove them

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,296 @@
+
+func TestAbortActionErrorCases(t *testing.T) {
+	tests := map[string]struct {
+		setupTransaction func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID)
```
t *testing.T is not needed in this function, it's not used anywhere

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,296 @@
+			setupTransaction: func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID) {
+				return "non-existent-reference", testusers.Alice.AuthID()
+			},
+			args: func(reference string) wdk.AbortActionArgs {
```
I'm almost sure that all those cases have the same implementation of args function

jakubmkowalski replied:
args removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,296 @@
+func TestAbortActionAbortableStatuses(t *testing.T) {
+	tests := map[string]struct {
+		setupTransaction func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID)
+		args             func(reference string) wdk.AbortActionArgs
```
args is not needed

jakubmkowalski replied:
removed

dorzepowski commented on pkg/storage/provider_abort_action_test.go:
```diff
@@ -0,0 +1,296 @@
+
+func TestAbortActionAbortableStatuses(t *testing.T) {
+	tests := map[string]struct {
+		setupTransaction func(t *testing.T, given testabilities.StorageFixture) (string, wdk.AuthID)
```
t *testing.T, is not needed

jakubmkowalski replied:
removed

dorzepowski commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -485,3 +554,55 @@ func (txs *Transactions) labelFilterScope(tx *gorm.DB, userID int, filter entity
+			{"update_transaction_status", func() error {
+				return txs.updateTransactionStatusByID(tx, transactionID, wdk.TxStatusFailed)
+			}},
+		}
```
### Not blocking comment, just how I see it.


This should be done differently, to not push the abortion logic to db layer.

We should
- read the whole Transaction with all it's relations as an entity,
- call the method Abort on that entity
- entity should remove and update all the fields that needs update
- pass entity to Save method or something like that on the repo
- and repo based on those changes should update db state.

Although it might look not optimal, it would show clearly the place where the abort logic is applied and what state is changed. Beside that, it shouldn't be the method that is often call, so having it suboptimal shouldn't be a problem.

jakubmkowalski commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func (a *abortAction) isPotentiallyTxID(reference string) bool {
+	return len(reference) == 64
```
changed

jakubmkowalski committed:
refactor: remove unused attributes and streamline test setup functions

dorzepowski commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -485,3 +554,55 @@ func (txs *Transactions) labelFilterScope(tx *gorm.DB, userID int, filter entity
 	}
 }
+
+func (txs *Transactions) AbortTransactionAtomic(ctx context.Context, transactionID uint, txID *string, reference string) error {
```
Please check this Issue from sonar: https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=395&id=bsv-blockchain_go-wallet-toolbox&open=AZg7TDpBiKNvHlIFi7PA

Maybe it's good to listen to it and make some refactoring to limit the cognition overhead.

jakubmkowalski replied:
refactored

dorzepowski commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -485,3 +554,55 @@ func (txs *Transactions) labelFilterScope(tx *gorm.DB, userID int, filter entity
+		historyBuilders := make([]history.Builder, len(txNotes))
+		for i, note := range txNotes {
+			historyBuilders[i] = history.NewBuilderFromNote(note)
+		}
```
Ohh, thanks to sonar, I just noticed that:
1. you create a builder
2. asked builder to create a note and create a slice with this one note
3. iterate over this slice with one note
4. create a builder from this one note
5. add a builder to a new slice

So it could be simplified and optimised by:
```suggestion
		historyBuilders := []history.Builder{history.NewBuilder().AbortAction(reference)}
```

jakubmkowalski replied:
simplified

dorzepowski commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -485,3 +554,55 @@ func (txs *Transactions) labelFilterScope(tx *gorm.DB, userID int, filter entity
+			if err := s.fn(); err != nil {
+				return fmt.Errorf("AbortTransactionAtomic: step '%s' failed: %w", s.name, err)
+			}
+		}
```
### optional suggestion

It could be changed to, and then those steps could be created once as a private var or in a separate function
```suggestion
		type step struct {
			name string
			fn   func(tx *gorm.DB, transactionID uint) error
		}

		steps := []step{
			{"release_reserved_utxos", txs.releaseReservedUTXOs},
			{"check_if_any_output_is_spent", txs.checkIfAnyOutputIsSpent},
			{"release_outputs_reserved", txs.releaseOutputsReservedByTransaction},
			{"delete_outputs", txs.deleteOutputsByTransactionID},
			{"update_transaction_status", func(tx *gorm.DB, transactionID uint) error {
				return txs.updateTransactionStatusByID(tx, transactionID, wdk.TxStatusFailed)
			}},
		}

		for _, s := range steps {
			if err := s.fn(tx, transactionID); err != nil {
				return fmt.Errorf("AbortTransactionAtomic: step '%s' failed: %w", s.name, err)
			}
		}
```

jakubmkowalski replied:
refactored

jakubmkowalski committed:
refactor: simplify transaction abort logic by using structured abort steps

jakubmkowalski committed:
style: add empty lines after returns

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

chris-4chain commented on pkg/storage/internal/actions/abort.go:
```diff
@@ -0,0 +1,78 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to find transaction by txid: %w", err)
+		}
+	}
```
I would make this error as:
`fmt.Errorf("failed to find transaction by txid %s: %w", args.Reference, err)`

And similarly, above, I would also print the reference in the error msg.

jakubmkowalski replied:
added context

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -206,23 +221,49 @@ func (txs *Transactions) FindTransactionByUserIDAndTxID(ctx context.Context, use
 			return nil, nil
 		}
+
 		return nil, fmt.Errorf("failed to find transaction by reference: %w", err)
```
AI Generated comment:
The `FindTransactionByReference` function has been changed to fetch a list of transactions but then checks if the list contains exactly one item. This approach can be inefficient and may hide a potential design issue. If the `reference` is intended to be unique per user, it would be better to enforce this with a unique constraint in the database. If multiple transactions can share the same reference, the function should be named `FindTransactionsByReference` to more accurately reflect its behavior. This will make the code clearer and prevent potential misunderstandings.

From actual KT:
@jakubmkowalski FYI: Find don't return `ErrRecordNotFound` on not found.
- But I would suggest to try to use the First item (that get's only one element)

jakubmkowalski replied:
changed to `First`, after changing `Reference` to unique index

jakubmkowalski committed:
refactor: update transaction reference field to unique index and simplify transaction retrieval logic

jakubmkowalski committed:
Merge branch 'feat/add_abort_action_storage' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_abort_action_storage

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

jakubmkowalski committed:
Merge branch 'main' into feat/add_abort_action_storage

chris-4chain commented on pkg/internal/fixtures/default_abort_action.go:
```diff
@@ -0,0 +1,17 @@
+)
+
+// DefaultValidAbortActionArgs returns default valid AbortActionArgs for testing
+func DefaultValidAbortActionArgs() wdk.AbortActionArgs {
```
This function seems unused - and since it's an internal thing, I would suggest to remove it.

jakubmkowalski replied:
removed

chris-4chain commented on pkg/internal/storage/repo/transactions.go:
```diff
@@ -207,26 +222,44 @@ func (txs *Transactions) FindTransactionByUserIDAndTxID(ctx context.Context, use
+	transaction := models.Transaction{}
+	err := txs.db.WithContext(ctx).
+		Scopes(scopes.UserID(userID)).
+		Where("tx_id = ?", txID).
```
Please add `index` on DB column: `txID`

jakubmkowalski replied:
added

jakubmkowalski committed:
fix: add index to TxID field in Transaction model for improved query performance

jakubmkowalski committed:
Merge branch 'feat/add_abort_action_storage' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_abort_action_storage
